asf_heat_pump_suitability/analysis/exploratory/create_full_dataset/assign_cluster_suitability_and_feasibility.py

```diff
@@ -2,236 +2,402 @@
+
+    # Compute not_in_listed_building
+    df = df.with_columns(
+        (~pl.col("in_listed_building")).alias("not_in_listed_building")
+    )
+    df = df.with_columns(
+        (~pl.col("in_conservation_area")).alias("not_in_conservation_area")
+    )
+
+    # Create has_outdoor_space from garden_area_m2
+    df = df.with_columns(
+        (pl.col("garden_area_m2") > outdoor_space_threshold).alias("has_outdoor_space")
+    )
+
+    # Creating close_to_city_centre from distance_to_city_centre and city_centre_threshold
+    df = df.with_columns(
+        pl.col("oa21").is_in(city_centre_oas).alias("close_to_city_centre")
+    )
+
+    # Creating imd_decile_above_avg from imd_decile
+    df = df.with_columns((pl.col("imd_decile") > 5).alias("imd_decile_above_avg"))
+
+    # Aggregating data by cluster, distance to anchor loads is calculated
+    # separately, so don't include this in this step
+
+    df = df.group_by("cluster").agg(
+        (
+            (pl.col([f for f in features if f != "close_to_anchor_loads"]).mean()).cast(
+                pl.Float64
             )
+            * 100
+        ).name.prefix("perc_"),
+        pl.col("UPRN").count().alias("cluster_size"),
+    )
+
+    # Creating close_to_anchor_loads from distance_from_anchor_property_m and anchor_loads_threshold
+
+    anchor_dist_df = anchor_dist_df.with_columns(
+        (pl.col("distance_from_anchor_property_m") <= anchor_loads_threshold).alias(
+            "close_to_anchor_loads"
         )
-        .alias("suitable_scheme")
+    ).select(
+        pl.col("cluster").cast(pl.Int64),
+        (pl.col("close_to_anchor_loads").cast(pl.Float64) * 100).name.prefix("perc_"),
     )
 
-    return cluster_df
+    df = df.join(anchor_dist_df, on="cluster")
+
+    # scale cluster size to be between 0 and 100
+    df = df.with_columns(
+        (
+            (pl.col("cluster_size") - pl.col("cluster_size").min())
+            / (pl.col("cluster_size").max() - pl.col("cluster_size").min())
+            * 100
+        ).alias("cluster_size")
+    )
+
+    return df
+
+
+def calculate_feasibility_expression(tech_specific_weights: dict) -> pl.Expr:
+    """
+    Generates a Polars expression to calculate a weighted feasibility score.
+
+    Args:
+        weights_dict: A dictionary of the format {feature_name: weight}.
+
+    Returns:
+        pl.Expr: A Polars expression for the calculated score.
+    """
+    # Sum of all weights to normalize the score
+    total_weight = sum(tech_specific_weights.values())
+
+    # Compute weighted sum i.e. sum(feature_value * weight)
+    weighted_cols_sum = pl.sum_horizontal(
+        (
+            pl.col(f"perc_{feature}") * weight
+            if feature != "cluster_size"  # most features will have a "perc_" prefix
+            else pl.col("cluster_size")
+            * weight  # cluster_size does not have a "perc_" prefix
+        )
+        for feature, weight in tech_specific_weights.items()
+    )
```

Comment:
I think you can do this but haven't tested (and then just return `weighted_cols_sum`)
```suggestion
    ).truediv(total_weight)
```